### PR TITLE
[mlir][amdgpu] Shared memory access optimization pass

### PR DESCRIPTION
**Pull request by erman-gurses**
It implements transformation to optimize accesses to shared memory.

Reference: https://reviews.llvm.org/D127457

_This change adds a transformation and pass to the NvGPU dialect that
attempts to optimize reads/writes from a memref representing GPU shared
memory in order to avoid bank conflicts. Given a value representing a
shared memory memref, it traverses all reads/writes within the parent op
and, subject to suitable conditions, rewrites all last dimension index
values such that element locations in the final (col) dimension are
given by newColIdx = col % vecSize + perm[row](col / vecSize, row)
where perm is a permutation function indexed by row and vecSize
is the vector access size in elements (currently assumes 128bit
vectorized accesses, but this can be made a parameter). This specific
transformation can help optimize typical distributed & vectorized accesses
common to loading matrix multiplication operands to/from shared memory._

### CODE REVIEW

**Comment by erman-gurses** [comment]
Hi Everyone, please let me know what you think about this PR. Thanks.

**Review by krzysz00 — COMMENTED** [review]
I'll have actual substantial comments later, but, as an initial matter, is a name as generic as "optimize shared memory access" appropriate here?

Also, I think, if you're trying to prevent aliasing issues around subviews, you might want to run a general alias analysis.

**Comment by erman-gurses** [comment]
> I'll have actual substantial comments later, but, as an initial matter, is a name as generic as "optimize shared memory access" appropriate here?

Can I get suggestion for that? I used this reference  https://reviews.llvm.org/D127457 for the implementation.

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
+//===- OptimizeSharedMemory.cpp - MLIR AMDGPU pass implementation
+//----------===//
```
Comment (arsenm) [inline mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:2]: Bad line break

File: mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPU.td [hunk 2]
```diff
@@ -29,6 +29,33 @@ def AMDGPU_Dialect : Dialect {
...
+
+    /// Defines the MemRef memory space attribute numeric value that indicates
+    /// a memref is located in global memory. This should correspond to the
+    /// value used in ROCDL.
+    static constexpr unsigned kGlobalMemoryAddressSpace = 1;
+
+    /// Defines the MemRef memory space attribute numeric value that indicates
+    /// a memref is located in private memory. This should correspond to the
+    /// value used in ROCDL.
+    static constexpr unsigned kPrivateMemoryAddressSpace = 2;
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPU.td:52]: Likewise...

File: mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp [hunk 3]
```diff
@@ -43,6 +43,21 @@ void AMDGPUDialect::initialize() {
       >();
 }
 
+bool amdgpu::AMDGPUDialect::isSharedMemoryAddressSpace(Attribute memorySpace) {
+  if (!memorySpace)
+    return false;
+  if (auto intAttr = llvm::dyn_cast<IntegerAttr>(memorySpace))
+    return intAttr.getInt() == AMDGPUDialect::kSharedMemoryAddressSpace;
```
Comment (joker-eph) [inline mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: Other than at the LLVM level, I wouldn't think we'll have integer based memory spaces.

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp (hunk 1 above)
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp]: Thanks.

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                                                    allocOp.getMemref())))
+        return;
+    }
+  }
+};
+} // namespace
+
+std::unique_ptr<Pass> mlir::amdgpu::createOptimizeSharedMemoryPass() {
+  return std::make_unique<OptimizeSharedMemoryPass>();
+}
```
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:251]: Regarding the three comments above including this one, When I removed those mentioned lines above, I am getting multiple errors. 
Starting with: `error: use of undeclared identifier 'createOptimizeSharedMemory'; did you mean 'registerOptimizeSharedMemory'?` 
Probably missing something.

[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:251]: Which line does this point to?

You may miss an include for the .h.inc in the header.

[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:251]: I added full error info below. Looks like in Passes.h.inc file Line at 204.

```
/llvm-project/build/tools/mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.h.inc:204:12: error: use of undeclared identifier 'createOptimizeSharedMemory'; did you mean 'registerOptimizeSharedMemory'?
    return createOptimizeSharedMemory();
           ^~~~~~~~~~~~~~~~~~~~~~~~~~
           registerOptimizeSharedMemory
/llvm-project/build/tools/mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.h.inc:202:13: note: 'registerOptimizeSharedMemory' declared here
inline void registerOptimizeSharedMemory() { ```

File: mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.h [hunk 5]
```diff
@@ -21,6 +21,10 @@ class ConversionTarget;
 namespace amdgpu {
 
 #define GEN_PASS_DECL_AMDGPUEMULATEATOMICSPASS
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.h:23]: Replace this macro with `GEN_PASS_DECL`

Comment (erman-gurses) [reply mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.h]: Thanks, that worked.

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  if (auto vectorReadOp = dyn_cast<vector::LoadOp>(op))
+    return vectorReadOp.getIndicesMutable().assign(indices);
+  if (auto vectorStoreOp = dyn_cast<vector::StoreOp>(op))
+    return vectorStoreOp.getIndicesMutable().assign(indices);
+  if (auto transferReadOp = dyn_cast<vector::TransferReadOp>(op))
+    return transferReadOp.getIndicesMutable().assign(indices);
+  if (auto transferWriteOp = dyn_cast<vector::TransferWriteOp>(op))
+    return transferWriteOp.getIndicesMutable().assign(indices);
+  llvm_unreachable("unsupported op type");
+}
```
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:48]: Sure, do you mean put them as polymorphic structure?

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+  OpBuilder builder(parentOp->getContext());
+
+  int64_t tgtDim = memRefType.getRank() - 1;
+  int64_t srcDim = memRefType.getRank() - 2;
+
+  // Transform indices for the ops writing to shared memory.
+  while (!shmWriteOps.empty()) {
+    Operation *shmWriteOp = shmWriteOps.back();
+    shmWriteOps.pop_back();
```
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp]: Thanks.

File: mlir/include/mlir/Dialect/AMDGPU/Transforms/Utils.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#include "mlir/IR/Operation.h"
+
+namespace mlir {
+namespace amdgpu {
+
+/// Get the indices that the given load/store operation is operating on.
+Operation::operand_range getIndices(Operation *op);
+
+/// Set the indices that the given load/store operation is operating on.
+void setIndices(Operation *op, ArrayRef<Value> indices);
```
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/include/mlir/Dialect/AMDGPU/Transforms/Utils.h:18]: Thanks, added that.

File: mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp (hunk 3 above)
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: Yeah, but roundtripping/backcompat/...

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+  return success();
+}
+
+namespace {
+class OptimizeSharedMemoryPass
+    : public amdgpu::impl::OptimizeSharedMemoryBase<OptimizeSharedMemoryPass> {
+public:
+  OptimizeSharedMemoryPass() = default;
+
+  void runOnOperation() override {
```
Comment (krzysz00) [inline mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:229]: The usual style these days seems to be moving these sorts of things out of the `namespace{}` block?

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    transformIndices(builder, shmReadOp->getLoc(), transformedIndices,
+                     memRefType, srcDim, tgtDim);
+    amdgpu::setIndices(shmReadOp, transformedIndices);
+  }
+
+  return success();
+}
+
+namespace {
+class OptimizeSharedMemoryPass
```
Comment (krzysz00) [inline mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:224]: Nit: usually a struct

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    return storeOp.getIndices();
+  if (auto vectorReadOp = dyn_cast<vector::LoadOp>(op))
+    return vectorReadOp.getIndices();
+  if (auto vectorStoreOp = dyn_cast<vector::StoreOp>(op))
+    return vectorStoreOp.getIndices();
+  if (auto transferReadOp = dyn_cast<vector::TransferReadOp>(op))
+    return transferReadOp.getIndices();
+  if (auto transferWriteOp = dyn_cast<vector::TransferWriteOp>(op))
+    return transferWriteOp.getIndices();
+  llvm_unreachable("unsupported op type");
```
[the remark this answers is not in the document]
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:31]: Yeah, i'd make this a `std::optional`

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:48]: Nah, as in, for example, `LoadOpInterface` and `StoreOpInterface`

File: mlir/test/Dialect/AMDGPU/optimize_shmem_reads_writes.mlir [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    // CHECK: [[c7:%.+]] = arith.constant 7 : index                     
+    // CHECK: [[srcBits:%.+]] = arith.andi [[fragRow]], [[c7]]          
+    // CHECK: [[c2:%.+]] = arith.constant 2 : index                     
+    // CHECK: [[xorBits:%.+]] = arith.shli [[srcBits]], [[c2]] 
+    // CHECK: [[fragColPerm:%.+]] = arith.xori [[fragCol:%.+]], [[xorBits]]
+    // CHECK:  vector.load [[shmA:%.+]][[[fragRow:%.+]], [[fragColPerm]]] : memref<128x32xf16, 3>, vector<8xf16>
+    %3 = vector.load %shmA[%fragRow, %fragColPerm] : memref<128x32xf16, 3>, vector<8xf16>
+    return
+  }
+
```
Comment (krzysz00) [inline mlir/test/Dialect/AMDGPU/optimize_shmem_reads_writes.mlir:57]: Nit: newline

File: mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.td [hunk 13]
```diff
@@ -30,4 +30,11 @@ def AmdgpuEmulateAtomicsPass : Pass<"amdgpu-emulate-atomics"> {
                         "Chipset that these operations will run on">];
 }
 
+def OptimizeSharedMemory : Pass<"amdgpu-optimize-shared-memory"> {
+  let summary = "Optimizes accesses to shared memory memrefs in order to reduce bank conflicts.";
```
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/include/mlir/Dialect/AMDGPU/Transforms/Passes.td:34]: Added, thanks.

File: mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp (hunk 9 above)
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/OptimizeSharedMemory.cpp:228]: Added that modification, removed the anonymous namespace

File: mlir/test/Dialect/AMDGPU/optimize_shmem_reads_writes.mlir (hunk 12 above)
Comment (erman-gurses) [reply mlir/test/Dialect/AMDGPU/optimize_shmem_reads_writes.mlir:57]: Don't we want extra empty line end of the test cases? I got some comments for my previous commits about it.

Comment (krzysz00) [reply mlir/test/Dialect/AMDGPU/optimize_shmem_reads_writes.mlir:57]: It's got the extra newline, yeah. I was noting you hadn't added one earlier

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp]: I see the point, I added optional.

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:47]: Thanks for the comments above. 

I did my homework for that. Like finding this interface here below and tried to design for implementation. 
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/MemRef/Transforms/RuntimeOpVerification.cpp#L351-L353

I did not figure out how to apply this example to our case (The interface already knows what concrete type is passed but in our case, we pass a generic _Operation_ type then figure out what concrete op type is in the function). Basically, it is not obvious how to use it. 

Do you have any pointers on that? If not, can we leave it like this?

File: mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp (hunk 3 above)
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: Thanks for the comments above. Krzysztof, that make sense. For backward compatibility reasons, can we leave like this?

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:47]: Please let me know when you can get back on this.

File: mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp (hunk 3 above)
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: Please let me know when you can get back on this.

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp:47]: You'd `dyn_cast` to the relevant `FooOpInterface` and then use the interface's methods.

File: mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp (hunk 3 above)
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: I'm fine with having the check for integral address spaces, if that helps

Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp:50]: Thanks.

**Comment by krzysz00** [comment]
@joker-eph I think we should land this even with the hardcoded list of memory operations because there's existing precedent, given that there's a followup plan for something like `MemRefAccessInterface` or the like.

File: mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (krzysz00) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp]: We could clean this up in the future, given that the NVVM equivalent of this pass already has a similar hardcoded function.

[the remark this answers is not in the document]
Comment (erman-gurses) [reply mlir/lib/Dialect/AMDGPU/Transforms/Utils.cpp]: Sounds good, definitely I am willing clean that up in the future with the interface version.

**Comment by joker-eph** [comment]
"existing precedent" isn't a reason IMO. It's not because a mistake slipped that it justified continuing a bad pattern...
But if @erman-gurses wants to tackle this as a separate PR that's fine by me.

**Comment by erman-gurses** [comment]
> "existing precedent" isn't a reason IMO. It's not because a mistake slipped that it justified continuing a bad pattern... But if @erman-gurses wants to tackle this as a separate PR that's fine by me.

@joker-eph, yes, I would like to tackle this as a separate PR.

**Comment by erman-gurses** [comment]
> "existing precedent" isn't a reason IMO. It's not because a mistake slipped that it justified continuing a bad pattern... But if @erman-gurses wants to tackle this as a separate PR that's fine by me.



> > "existing precedent" isn't a reason IMO. It's not because a mistake slipped that it justified continuing a bad pattern... But if @erman-gurses wants to tackle this as a separate PR that's fine by me.
> 
> @joker-eph, yes, I would like to tackle this as a separate PR.

and @joker-eph, please let me know when you can approve this.

**Comment by joker-eph** [comment]
Probably best if someone more involved with AMDGPU reviews and approves?

**Comment by erman-gurses** [comment]
> Probably best if someone more involved with AMDGPU reviews and approves?

Sounds good, @krzysz00 has already reviewed it. Krzysztof, let me know when you can approve this. Thanks.

**Review by krzysz00 — APPROVED** [review]
This looks reasonable and we'll clean this and its NVPTX equivalent up in a followup, it sounds like.

**Comment by joker-eph** [comment]
Sorry, had to revert because of the bot breakage. Please see: https://lab.llvm.org/buildbot/#/builders/61/builds/53218 and feel free to reland when you have a fix!

**Comment by erman-gurses** [comment]
> Sorry, had to revert because of the bot breakage. Please see: https://lab.llvm.org/buildbot/#/builders/61/builds/53218 and feel free to reland when you have a fix!

Sure, I will investigate and reland later.

**Pull request merged**